fdbserver/ClusterController.actor.cpp

```diff
@@ -3295,108 +3369,165 @@ void checkOutstandingStorageRequests(ClusterControllerData* self) {
+	if (self->db.config.isExcludedServer(worker.interf.addresses())) {
+		bestFitness = std::max(bestFitness, ProcessClass::ExcludeFit);
+	}
+	return bestFitness;
+}
+
+// Returns true iff the singleton is healthy. "Healthy" here means that
+// the singleton is stable (see below) and doesn't need to be rerecruited.
+// Side effects: (possibly) initiates recruitment
+template <class Interface>
+bool isHealthySingleton(ClusterControllerData* self,
+                        const WorkerDetails& newWorker,
+                        const Singleton<Interface>& singleton,
+                        const ProcessClass::Fitness& bestFitness,
+                        const Optional<UID> recruitingID) {
+	// A singleton is stable if it exists in cluster, has not been killed off of proc and is not being recruited
+	bool isStableSingleton = singleton.interface.present() &&
+	                         self->id_worker.count(singleton.interface.get().locality.processId()) &&
+	                         (!recruitingID.present() || (recruitingID.get() == singleton.interface.get().id()));
+
+	if (!isStableSingleton) {
+		return false; // not healthy because unstable
+	}
+
+	auto& currWorker = self->id_worker[singleton.interface.get().locality.processId()];
+	auto currFitness = currWorker.details.processClass.machineClassFitness(singleton.getClusterRole());
+	if (currWorker.priorityInfo.isExcluded) {
+		currFitness = ProcessClass::ExcludeFit;
+	}
+	// If any of the following conditions are met, we will switch the singleton's process:
+	// - if the current proc is used by some non-master, non-singleton role
+	// - if the current fitness is less than optimal (lower fitness is better)
+	// - if currently at peak fitness but on same process as master, and the new worker is on different process
+	bool shouldRerecruit =
+	    self->isUsedNotMaster(currWorker.details.interf.locality.processId()) || bestFitness < currFitness ||
+	    (currFitness == bestFitness && currWorker.details.interf.locality.processId() == self->masterProcessId &&
+	     newWorker.interf.locality.processId() != self->masterProcessId);
+	if (shouldRerecruit) {
+		std::string roleAbbr = singleton.getRole().abbreviation;
+		TraceEvent(("CCHalt" + roleAbbr).c_str(), self->id)
+		    .detail(roleAbbr + "ID", singleton.interface.get().id())
+		    .detail("Excluded", currWorker.priorityInfo.isExcluded)
+		    .detail("Fitness", currFitness)
+		    .detail("BestFitness", bestFitness);
+		singleton.recruit(self); // SIDE EFFECT: initiating recruitment
+		return false; // not healthy since needed to be rerecruited
+	} else {
+		return true; // healthy because doesn't need to be rerecruited
+	}
+}
+
+// Returns a mapping from pid->pidCount for pids
+std::map<Optional<Standalone<StringRef>>, int> getColocCounts(const vector<Optional<Standalone<StringRef>>>& pids) {
+	std::map<Optional<Standalone<StringRef>>, int> counts;
+	for (const auto& pid : pids) {
+		if (counts.find(pid) == counts.end()) {
```

Comment:
count will get initialized to 0 when using [], so you can just do ++counts[pid]; without the find